crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -3,74 +3,172 @@ use containerd_shim_wasm::container::{
 }
 
+impl Default for WasmtimeEngine {
+    fn default() -> Self {
+        let mut config = wasmtime::Config::new();
+        config.wasm_component_model(true);
+        Self {
+            engine: wasmtime::Engine::new(&config)
+                .context("failed to create wasmtime engine")
+                .unwrap(),
+        }
+    }
+}
+
+pub struct Data {
+    wasi: wasi_preview2::WasiCtx,
+    table: Table,
+}
+
+impl wasmtime_wasi::preview2::WasiView for Data {
+    fn table(&self) -> &Table {
+        &self.table
+    }
+
+    fn table_mut(&mut self) -> &mut Table {
+        &mut self.table
+    }
+
+    fn ctx(&self) -> &wasi_preview2::WasiCtx {
+        &self.wasi
+    }
+
+    fn ctx_mut(&mut self) -> &mut wasi_preview2::WasiCtx {
+        &mut self.wasi
+    }
+}
+
+pub enum WasmBinaryType {
+    Module,
+    Component,
+}
+
+fn wasm_binary_type(bytes: &[u8]) -> Option<WasmBinaryType> {
+    if bytes.starts_with(b"\0asm\x01\0\0\0") {
+        return Some(WasmBinaryType::Module);
+    }
+    if bytes.starts_with(b"\0asm\x0d\0\x01\0") {
+        return Some(WasmBinaryType::Component);
+    }
+    None
+}
+
 impl Engine for WasmtimeEngine {
     fn name() -> &'static str {
         "wasmtime"
     }
 
     fn run_wasi(&self, ctx: &impl RuntimeContext, stdio: Stdio) -> Result<i32> {
         log::info!("setting up wasi");
-        let path = Dir::from_std_file(std::fs::File::open("/")?);
+        let root_path = Dir::from_std_file(std::fs::File::open("/")?);
         let envs: Vec<_> = std::env::vars().collect();
-
-        let mut wasi_builder = WasiCtxBuilder::new();
-        wasi_builder
-            .args(ctx.args())?
-            .envs(envs.as_slice())?
-            .inherit_stdio()
-            .preopened_dir(path, "/")?;
-
-        stdio.redirect()?;
-
-        log::info!("building wasi context");
-        let wctx = wasi_builder.build();
-
-        log::info!("wasi context ready");
         let WasiEntrypoint { path, func } = ctx.wasi_entrypoint();
 
-        let module = match ctx.wasm_layers() {
+        let data = match ctx.wasm_layers() {
             [] => {
                 log::info!("loading module from file");
                 let path = path
                     .resolve_in_path_or_cwd()
                     .next()
                     .context("module not found")?;
-
-                Module::from_file(&self.engine, path)?
+                std::fs::read(path)?
             }
             [module] => {
                 log::info!("loading module wasm OCI layers");
-                Module::from_binary(&self.engine, &module.layer)?
+                module.layer.clone()
             }
             [..] => bail!("only a single module is supported when using images with OCI layers"),
         };
 
-        let mut linker = Linker::new(&self.engine);
+        let status = match wasm_binary_type(&data) {
+            Some(WasmBinaryType::Module) => {
+                let module = Module::from_binary(&self.engine, &data)?;
```

Comment:
would it make sense from a readability/maintaince to move these longer sections calls to functions?